ast/build.rs

```diff
@@ -0,0 +1,130 @@
+    //     import asdl_rs
+    //     asdl_rs.main(
+    //         "./Python.asdl",
+    //         "{mod_path}",
+    //         "{def_path}",
+    //     )"#);
+    //     py.eval(&code, None, None)?.extract()
+    // }).expect("failed");
+
+    let mut cmd = std::process::Command::new("python3");
+    let output = cmd
+        .args([
+            "./asdl_rs.py",
+            "-D",
+            &def_path,
+            "-M",
+            &mod_path,
+            "./Python.asdl",
+        ])
+        .output();
+    let success = output.as_ref().map_or(false, |o| o.status.success());
+    if success {
+        // failing fmt is ok
+        let _ = std::process::Command::new("rustfmt")
+            .args([&def_path, &mod_path])
+            .output();
+        let c1 = file_changed(&def_path, "bootstrap/ast_def.rs").unwrap();
+        let c2 = file_changed(&mod_path, "bootstrap/ast_mod.rs").unwrap();
+        if c1 || c2 {
+            // ensure bootstrap files are latest
+            let _ = std::fs::copy(def_path, "bootstrap/ast_def.rs");
+            let _ = std::fs::copy(mod_path, "bootstrap/ast_mod.rs");
+        }
+    } else {
+        let mut error = match output {
+            Ok(out) => std::str::from_utf8(out.stderr.as_slice())
+                .expect("stderr is not utf8")
+                .to_owned(),
+            Err(e) => format!("{cmd:?}\n{e:?}"),
+        };
+        error.insert(0, '\n');
+        let error = error.replace('\n', "\n//    ");
+        fn bootstrap(
+            source: impl AsRef<Path>,
+            target: impl AsRef<Path>,
+            msg: &str,
+        ) -> std::io::Result<()> {
+            let content = read_file_content(source)?;
+            let mut file = File::create(target)?;
+            writeln!(file, "// NEVER COMMIT THIS FILE TO REPOSITORY!")?;
+            writeln!(
+                file,
+                "// This content copied from `ast/bootstrap/` because generation failed."
+            )?;
+            writeln!(file, "// To prevent bootstrapping, build RustPython then copy `rustpython` binary into executable path as name of `python3`")?;
+            writeln!(file, "// Error:{msg}")?;
+            writeln!(file, "//")?;
+            file.write_all(content.as_bytes())?;
+            Ok(())
+        }
+        bootstrap("bootstrap/ast_def.rs", "src/ast_gen.rs", &error)
+            .expect("failed to copy bootstrap ast_def.rs");
+        bootstrap(
+            "bootstrap/ast_mod.rs",
+            "../vm/src/stdlib/ast/gen.rs",
+            &error,
+        )
+        .expect("failed to copy bootstrap ast_mod.rs");
+    }
+}
+
+fn read_file_content(path: impl AsRef<Path>) -> std::io::Result<String> {
+    let path = path.as_ref();
+    let mut file = File::open(path)?;
+    let mut content = String::new();
+    file.read_to_string(&mut content)?;
+    Ok(content)
+}
+
+fn file_changed(path1: impl AsRef<Path>, path2: impl AsRef<Path>) -> std::io::Result<bool> {
```

Comment:
mtime isn't as reliable as something like a checksum; I'm pretty sure git checkout touches the mtime and ideally this would only be run when necessary (i.e. in development, not when building as a dependency or something)

Reply:
this function does not use mtime. it actually read the file content and compare.